### PR TITLE
Snapshot results on iterating

### LINKED ISSUE #166: Snapshot RealmResults before iteration
Change _RealmResultsIterator implementation to make a snapshot of results before starting iteration.
Currently blocked by realm-core issues [#5154](https://github.com/realm/realm-core/issues/5154)

### PR DESCRIPTION
**Pull request by desistefanova**
Snapshot results of RealmObjects when iterating. Fixes #166.

### CODE REVIEW

File: lib/src/results.dart [hunk 1]
```diff
@@ -52,7 +51,15 @@ class RealmResults<T extends RealmObject> extends collection.IterableBase<T> {
 
   /// Returns a new `Iterator` that allows iterating the elements in this `RealmResults`
   @override
-  Iterator<T> get iterator => _RealmResultsIterator(this);
+  Iterator<T> get iterator {
+    RealmResults<T> results = this;
+    bool isSubtypeOfRealmObject = (<T>[] is List<RealmObject?>);
```
Comment (desistefanova) [inline lib/src/results.dart:56]:
```suggestion
    bool isSubtypeOfRealmObject = (<T>[] is List<RealmObject>);
```

File: lib/src/results.dart [hunk 2]
```diff
@@ -56,7 +58,14 @@ class RealmResults<T extends RealmObject> extends collection.IterableBase<T> {
 
   /// Returns a new `Iterator` that allows iterating the elements in this `RealmResults`.
   @override
-  Iterator<T> get iterator => _RealmResultsIterator(this);
+  Iterator<T> get iterator {
+    RealmResults<T> results = this;
```
Comment (nielsenko) [inline lib/src/results.dart:62]:
```suggestion
    var results = this;
```

File: test/results_test.dart [hunk 3]
```diff
@@ -138,7 +138,7 @@ Future<void> main([List<String>? args]) async {
     realm.close();
   });
 
-  test('Results iteration test', () {
+  test('Results iterating test', () {
```
Comment (nielsenko) [inline test/results_test.dart:141]: I think _iteration_ is the better word here, but then again english is not my native tongue.

**Review by nielsenko — APPROVED** [review]
A few minor things..

File: CHANGELOG.md [hunk 4]
```diff
@@ -16,6 +16,9 @@ x.x.x Release notes (yyyy-MM-dd)
 ### Compatibility
 * Dart ^2.15 on Windows, MacOS and Linux
 
+### Fixed
+* Iterating RealmResult will snapshot the collection if the items are RealmObject from Realm. ([#258](https://github.com/realm/realm-dart/pull/258))
```
Comment (blagoev) [inline CHANGELOG.md:20]:
```suggestion
* Snapshot the results collection when Iterating if the items are realm objects. ([#258](https://github.com/realm/realm-dart/pull/258))
```

**Review by blagoev — APPROVED** [review]
added some suggestions

File: test/results_test.dart (hunk 3 above)
Comment (desistefanova) [reply test/results_test.dart:141]: Done

File: CHANGELOG.md [hunk 5]
```diff
@@ -16,6 +16,9 @@ x.x.x Release notes (yyyy-MM-dd)
 ### Compatibility
 * Dart ^2.15 on Windows, MacOS and Linux
 
+### Fixed
+* Snapshot the results collection when Iterating if the items are realm objects. ([#258](https://github.com/realm/realm-dart/pull/258))
```
Comment (nielsenko) [inline CHANGELOG.md:20]:
```suggestion
* Snapshot the results collection when iterating if the items are realm objects. ([#258](https://github.com/realm/realm-dart/pull/258))
```

**Review by nielsenko — APPROVED** [review]

**Review by blagoev — APPROVED** [review]

**Pull request merged**